### PR TITLE
Mapped type and string index signature relations

### LINKED ISSUE #14548: Mapped types over keyof T should be subtypes of types with string index signatures
**TypeScript Version:**  2.2.1

**Code**

```ts
// helpers to map types
type Dictionary = {
    [key: string]: string;
};

interface IBaseEntity {
    name: string;
    properties: Dictionary;
}
interface IEntity<t> extends IBaseEntity {
    properties: Record<t, string>;
}
```

**Expected behaviour:**
No Compile error.

**Actual behaviour:**
Compile error: 
```
Interface 'IEntity<t>' incorrectly extends interface 'IBaseEntity'.
  Types of property 'properties' are incompatible.
    Type 'Record<t, string>' is not assignable to type 'Dictionary'.
```

Edit: Replace StrictDictionary<t> with Record<t, string>. Show full compile error.

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this PR a mapped type `{ [P in K]: T }`, where `K` is generic, is related to a string index signature `{ [key: string]: U }` if `T` is related to `U`. For example:

```ts
function f<K extends string>(x: { [key: string]: number }, y: Record<K, number>) {
    x = y;  // Ok
}
```
Fixes #14548. Also fixes an unintended effect of #17382 that would allow *any* generic mapped type to be assignable to a string index signature (because generic mapped types have no manifest properties they were being mistaken for empty object types).

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@rbuckton @sandersn I'd like to get this one into 2.5 RC so take a look.

File: tests/cases/compiler/indexSignatureAndMappedType.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+// A mapped type { [P in K]: X }, where K is a generic type, is related to
+// { [key: string]: Y } if X is related to Y.
+
+function f1<T, K extends string>(x: { [key: string]: T }, y: Record<K, T>) {
+    x = y;
+    y = x;  // Error
+}
+
+function f2<T, K extends string>(x: { [key: string]: T }, y: Record<string, T>) {
```
Comment (sandersn) [inline tests/cases/compiler/indexSignatureAndMappedType.ts]: what's the point of this test? `K` isn't used.

Comment (sandersn) [inline tests/cases/compiler/indexSignatureAndMappedType.ts]: I guess it's to assert that the key type of `Record` doesn't matter, only the mapped type's template type. Seems self-evident to me, but maybe it's worth a test.

Comment (rbuckton) [inline tests/cases/compiler/indexSignatureAndMappedType.ts]: To me it appears to be unused and unnecessary for the test. @ahejlsberg, is the type parameter `K` needed for this test?

Comment (ahejlsberg) [inline tests/cases/compiler/indexSignatureAndMappedType.ts]: Yeah, no need to have that type parameter. Will fix.

**Pull request merged**